packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EditingContextCrossReferenceAdapter.java

```diff
@@ -0,0 +1,121 @@
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.EStructuralFeature.Setting;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.util.ECrossReferenceAdapter;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+
+/**
+ * A {@link ECrossReferenceAdapter}@ that is able to clean dangling references when an object is deleted or when a
+ * resource is removed from the ResourceSet.
+ *
+ * @author lfasani
+ */
+public class EditingContextCrossReferenceAdapter extends ECrossReferenceAdapter {
+
+    @Override
+    protected void handleContainment(Notification notification) {
+        Object oldValue = notification.getOldValue();
+        int eventType = notification.getEventType();
+
+        if (eventType == Notification.REMOVE || eventType == Notification.REMOVE_MANY) {
+            if (eventType == Notification.REMOVE) {
+                this.handleRemoveObject(oldValue);
+
+            } else if (eventType == Notification.REMOVE_MANY) {
+                for (Object object : (Collection<?>) oldValue) {
+                    this.handleRemoveObject(object);
+                }
+            }
+        } else {
+            super.handleContainment(notification);
+        }
+    }
+
+    private void handleRemoveObject(Object object) {
+        if (object instanceof EObject) {
+            this.handleRemoveEObject((EObject) object);
+        } else if (object instanceof Resource) {
+            this.handleRemoveResource((Resource) object);
+        }
+    }
+
+    private void handleRemoveEObject(EObject eObject) {
+        // Remove all inverse references
+        this.removeProxyFromInverseReference(eObject);
+        TreeIterator<Object> allProperContents = EcoreUtil.getAllProperContents(eObject, false);
+        while (allProperContents.hasNext()) {
+            Object object = allProperContents.next();
+            if (object instanceof EObject) {
+                this.removeProxyFromInverseReference((EObject) object);
+            }
+        }
+
+        // clean the crossRefererenceAdapter
+        this.unsetTarget(eObject);
+    }
+
+    private void handleRemoveResource(Resource resource) {
+        // Remove all inverse references
+        TreeIterator<Object> allProperContents = EcoreUtil.getAllProperContents(resource, false);
+        while (allProperContents.hasNext()) {
+            Object object = allProperContents.next();
+            if (object instanceof EObject) {
+                this.removeProxyFromInverseReference((EObject) object);
+            }
+        }
+
+        // clean the crossRefererenceAdapter
+        this.unsetTarget(resource);
+    }
+
+    private void removeProxyFromInverseReference(EObject referencedObject) {
+        Collection<Setting> settings = this.getInverseReferences(referencedObject);
+
+        if (settings != null) {
+            Collection<Setting> nonContainmentReferences = this.getNonContainmentReferences(settings);
```

Comment:
This method (`removeProxyFromInverseReference`) will be called *a lot*. It should be as efficient as possible. It should do the iteration and test on the references itself instead of calling `getNonContainmentReferences` which allocates a new `ArrayList` every time.
Maybe it should even avoid calling `getInverseReferences` (which does quite a lot of work, some which may not be needed in our case) and use `ECrossReferenceAdapter.inverseCrossReferencer` directly.

Reply:
I tried to `ECrossReferenceAdapter.inverseCrossReferencer` directly but it fails for the `EOpposite` features.
It is actually, the aim of `getNonContainmentReferences ` to manage the EOpposite